3 course/OperatingSystem/mythread.c

```diff
@@ -0,0 +1,173 @@
+            return;
+        }
+
+        err = syscall(SYS_futex, &lock, FUTEX_WAIT, 0, NULL, NULL, 0);
+        if (err == -1 && errno != EAGAIN) {
+            printf("Futex FUTEX_WAIT failed: %s\n", strerror(errno));
+            abort();
+        }
+    }
+}
+
+void wake_mythread(volatile int* lock) {
+    const int zero = 0;
+    int err; 
+   
+    if (atomic_compare_exchange_strong(lock, &zero, 1)) {
+        err = syscall(SYS_futex, &lock, FUTEX_WAKE, 1, NULL, NULL, 0);
+        if (err == -1) {
+            printf("Futex FUTEX_WAKE failed: %s\n", strerror(errno));
+            abort();
+        }
+    }
+}
+
+int mythread_startup(void* arg) {
+    mythread_struct_t* mythread = (mythread_struct_t*)arg;
+
+    mythread->retval = mythread ->start_routine(mythread->arg);
+    //mythread->exited = 1;
+    //меняем значение в функции ниже
+    wake_mythread(&mythread->exited);
+
+    //wait until join (тот самый futex!)
+    wait_mythread(&mythread->joined);
+
+    return 0;
+}
+
+void destroy(mythread_struct_t* mythread) {
+    munmap((void*)mythread, STACK_SIZE);
+}
+
+int mythread_join(uintptr_t mytid, void** retval) {
+    mythread_struct_t* mythread = (mythread_struct_t*)mytid;
+
+    //wait until thread ends
+    wait_mythread(&mythread->exited);
+
+    printf("thread join: the thread %d finished\n", mythread->mythread_id);
+
+    *retval = mythread->retval;
+
+    //mythread->joined = 1;
+    //меняю значение в функции ниже
+    wake_mythread(&mythread->joined);
+
+    destroy(mythread);
+
+    return 0;
+}
+
+int mythread_create(uintptr_t* mytid, start_routine_t start_routine, void* arg) {
+    static int thread_num = 0;
+    mythread_struct_t* mythread;
+    int child_pid;
+    void* child_stack;
+
+    thread_num++;
+
+    printf("mythread_create: creating thread %d\n", thread_num);
+
+    child_stack = create_stack(STACK_SIZE, thread_num);
+    mprotect(child_stack + sizeof(mythread_struct_t), PAGE, PROT_NONE);
+
+    mythread = (mythread_struct_t*)(child_stack + sizeof(mythread_struct_t));
+    mythread->mythread_id = thread_num;
+    mythread->start_routine = start_routine;
+    mythread->arg = arg;
+    mythread->retval = NULL;
+    mythread->joined = 0;
+    mythread->exited = 0;
+
+    child_pid = clone(mythread_startup, child_stack + STACK_SIZE, CLONE_VM | CLONE_FILES | CLONE_SYSVSEM|  CLONE_THREAD | CLONE_SIGHAND | SIGCHLD, (void*)mythread);
+    if (child_pid == -1) {
+        printf("clone failed: %s\n", strerror(errno));
+        exit(-1);
+    }
+
+    *mytid = (uintptr_t)mythread;
+
+    return 0;
+}
+
+void* mythread_func (void* arg) {
+    char* str = (char*)arg;
+    for(int i = 0; i < 5; i++) {
+        printf("hello:\n %s\n", str);
+        sleep(1);
+    }
+    return (void*)"solnishko luchistoe lubit skakat`";
+}
+
+int main() {
+    uintptr_t tid;
+    void* retval;
+
+    printf("main [%d %d %d]\n", getpid(), getppid(), gettid());
+
+    mythread_create(&tid, mythread_func, "hello from main");
+    sleep(10);
```

Comment:
why do we need sleep here ? Looks like it hides some possible problems.

Reply:
This is for debag

Reply:
please remove and check if it works fine